compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/PromisedValue.kt

```diff
@@ -70,8 +78,32 @@ abstract class PromisedValue(val codegen: ExpressionCodegen, val type: Type, val
 
     val typeMapper: IrTypeMapper
         get() = codegen.typeMapper
-}
 
+    private fun IrType.isJava() =
+        (this.classifierOrFail.owner as IrDeclaration).origin == IrDeclarationOrigin.IR_EXTERNAL_JAVA_DECLARATION_STUB
+
+    private fun IrType.isPackagePrivate(): Boolean {
+        return this.getClass()?.visibility?.delegate == JavaVisibilities.PackageVisibility
+    }
+
+    private fun IrType.inDifferentPackage(): Boolean =
+        this.getClass()!!.getPackageFragment()?.fqName != codegen.irFunction.getPackageFragment()?.fqName
+
+    private fun computeCoerceTarget(
+        isSubType: Boolean,
+        erasedTargetType: IrType,
+        default: Type
+    ): Type {
+        if (!isSubType && erasedTargetType.isPackagePrivate() && erasedTargetType.inDifferentPackage()) {
+            val computed = irType.superTypes().firstOrNull {
```

Comment:
Please, check what happens in the following tests (should be an error in both cases):
```
// FILE: test.kt
import j.J
import j.PPImpl1
import j.PPImpl2

fun <T> select(a: T, b: T) = a

fun box(): String =
    J.test(
        select(PPImpl1(), PPImpl2())
    )

// FILE: j/J.java
package j;

public class J {
    public static String test(PP pp) {
        return pp.test();
    }
}

// FILE: j/PA.java
package j;

public interface PA {
}

// FILE: j/PP.java
package j;

interface PP {
    String test();
}

// FILE: j/PPImpl1.java
package j;

public class PPImpl1 implements PA, PP {
    @Override
    public String test() {
        return "OK";
    }
}

// FILE: j/PPImpl2.java
package j;

public class PPImpl2 implements PA, PP {
    @Override
    public String test() {
        return "OK";
    }
}
```


```
// FILE: test.kt
import j.J
import j.PPImpl1
import j.PPImpl2

fun <T> select(a: T, b: T) = a

fun box(): String =
    J.test(
        select(PPImpl1(), PPImpl2())
    )

// FILE: j/J.java
package j;

public class J {
    public static String test(PP pp) {
        return pp.test();
    }
}

// FILE: j/PQ.java
package j;

public interface PQ {
}

// FILE: j/PP.java
package j;

interface PP {
    String test();
}

// FILE: j/PPImpl1.java
package j;

public class PPImpl1 implements PP, PQ {
    @Override
    public String test() {
        return "OK";
    }
}

// FILE: j/PPImpl2.java
package j;

public class PPImpl2 implements PP, PQ {
    @Override
    public String test() {
        return "OK";
    }
}
```

This is to make sure that we handle approximated intersection types properly.

Reply:
An error produced in both cases, I've add them to the tests.